src/main/java/org/jenkinsci/test/acceptance/po/PluginManager.java

```diff
@@ -186,101 +197,91 @@ public boolean installPlugins(final String... specs) throws UnableToResolveDepen
             if (!someChangeRequired) {
                 return false;
             }
-            List<PluginMetadata> pluginToBeInstalled = ucmd.get().transitiveDependenciesOf(jenkins, candidates);
-            for (PluginMetadata newPlugin : pluginToBeInstalled) {
+            List<PluginMetadata> pluginToBeInstalled = ucmd.get().transitiveDependenciesOf(jenkins, Arrays.asList(specs));
+            for (PluginMetadata newPlugin: pluginToBeInstalled) {
                 final String name = newPlugin.getName();
-                String claimedVersion = candidates.get(name);
+                String requiredVersion = candidates.get(name);
                 String availableVersion = newPlugin.getVersion();
-                if (claimedVersion == null) { // a dependency
-                    claimedVersion = availableVersion;
+                if (requiredVersion == null) { // a dependency
+                    requiredVersion = availableVersion;
                 }
-                final String currentSpec = StringUtils.isNotEmpty(claimedVersion)
-                    ? name + "@" + claimedVersion
+                final String currentSpec = StringUtils.isNotEmpty(requiredVersion)
+                    ? name + "@" + requiredVersion
                     : name
                 ;
                 InstallationStatus status = installationStatus(currentSpec);
                 if (status != InstallationStatus.UP_TO_DATE) {
+                    if (new VersionNumber(requiredVersion).compareTo(new VersionNumber(availableVersion)) > 0) {
+                        throw new AssumptionViolatedException(
+                                name + " has version " + availableVersion + " but " + requiredVersion + " was requested");
+                    }
+                    File localFile = newPlugin.resolve(injector, availableVersion);
                     try {
-                        if (new VersionNumber(claimedVersion).compareTo(new VersionNumber(availableVersion)) > 0) {
-                            throw new AssumptionViolatedException(
-                                    name + " has version " + availableVersion + " but " + claimedVersion + " was requested");
-                        }
-                        newPlugin.uploadTo(jenkins, injector, availableVersion);
```

Comment:
The old method was fast—a direct POST. Now we are going through the web driver to visit `/pluginManager/advanced` and upload via HTML, which is a lot slower. Is there no middle ground?